pretyping/vnorm.ml

```diff
@@ -78,9 +78,9 @@ let type_constructor mind mib u (ctx, typ) params =
 
 
 
-let construct_of_constr const env tag typ =
+let construct_of_constr const env sigma tag typ =
   let (t, allargs) = decompose_appvect (whd_all env typ) in
-  match Constr.kind t with
+  match EConstr.kind_upto sigma t with
```

Comment:
actually we should probably be changing the whd_all call too
Let me try to get an example

Reply:
`Definition barvm (z:nat) (t:=_ z) (x := true : t) := Eval vm_compute in x.` still failing

Reply:
The native code does
```ocaml
let t = whd_all env c in
let (t, allargs) = try destApp t with DestKO -> (t,[||])
```
Is that the way to go?

Reply:
AFAICT no, it needs to be changed to evar-aware whd_all too.
Does the example work with native compute?

Reply:
Right, it fails too.

Reply:
So, `whd_all env sigma`.

What do you make of the `destApp` dance? Do I add it to `vm_compute`? Remove it from `native_compute`? Neither?

Reply:
Sorry, that was a dumb question. It is the same code both sides, just written in a different way.

Reply:
decompose_appvect is a nicer way to do the same thing 
Do what you like with it.

Reply:
I am a bit worried by the sheer amount of `whd_all` calls in those files. Do they all need to be converted? Or are some of them correct? (You can also tell me that it does not hurt to convert them all.)

Reply:
I don't know.

Reply:
I went for the minimal change. (No idea if there is a more idiomatic way of writing it.) Let us see how it fares.

Reply:
We don't need kind_upto as whd_all will handle it btw

Reply:
This is getting ridiculous. File `test-suite/bugs/closed/bug_5096.v`, which is supposed to terminate instantly, now takes 20 seconds.

Reply:
So, if I understand #5096 correctly, we are not allowed to use `Reductionops.whd_all` here, because it is way too slow.

Reply:
Probably
cc @ppedrot

Reply:
You may try to use `clos_whd_flags` instead.

Reply:
Seems to work better. (What is the difference?) Let us see if it succeeds this time.